src/mpsc/mod.rs

```diff
@@ -0,0 +1,182 @@
+    shared.closed = true;
+  }
+}
+
+impl<T> Receiver<T> {
+  /// Receives a value from the channel, returning `None` if there
+  /// are no more items and the channel is closed.
+  pub async fn recv(&mut self) -> Option<T> {
+    // note: this is `&mut self` so that it can't be polled
+    // concurrently. DO NOT change this to `&self` because
+    // then futures will lose their wakers.
+    RecvFuture {
+      shared: &self.shared,
+    }
+    .await
+  }
+
+  /// Number of pending unread items.
+  pub fn len(&self) -> usize {
+    self.shared.borrow().queue.len()
+  }
+
+  /// If the receiver has no pending items.
+  pub fn is_empty(&self) -> bool {
+    self.len() == 0
+  }
+}
+
+struct RecvFuture<'a, T> {
+  shared: &'a RefCell<Shared<T>>,
+}
+
+impl<'a, T> Future for RecvFuture<'a, T> {
+  type Output = Option<T>;
+
+  fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+    let mut shared = self.shared.borrow_mut();
+    if let Some(value) = shared.queue.pop_front() {
+      Poll::Ready(Some(value))
+    } else if shared.closed {
+      Poll::Ready(None)
+    } else {
+      shared.waker.register(cx.waker());
+      Poll::Pending
+    }
+  }
+}
+
+struct Shared<T> {
+  queue: ChunkedQueue<T>,
+  waker: UnsyncWaker,
+  closed: bool,
+}
+
+/// A ![`Sync`] and ![`Sync`] equivalent to `tokio::sync::unbounded_channel`.
+pub fn unbounded_channel<T>() -> (Sender<T>, Receiver<T>) {
+  let shared = Rc::new(RefCell::new(Shared {
+    queue: ChunkedQueue::default(),
+    waker: UnsyncWaker::default(),
+    closed: false,
+  }));
+  (
+    Sender {
+      shared: shared.clone(),
+    },
+    Receiver { shared },
+  )
+}
+
+#[cfg(test)]
+mod test {
+  use tokio::join;
+
+  use super::*;
+
+  #[tokio::test(flavor = "current_thread")]
+  async fn sends_receives_exits() {
+    let (sender, mut receiver) = unbounded_channel::<usize>();
+    sender.send(1).unwrap();
+    assert_eq!(receiver.recv().await, Some(1));
+    sender.send(2).unwrap();
+    assert_eq!(receiver.recv().await, Some(2));
+    drop(sender);
+    assert_eq!(receiver.recv().await, None);
+  }
+
+  #[tokio::test(flavor = "current_thread")]
+  async fn sends_multiple_then_drop() {
+    let (sender, mut receiver) = unbounded_channel::<usize>();
+    sender.send(1).unwrap();
+    sender.send(2).unwrap();
+    drop(sender);
+    assert_eq!(receiver.len(), 2);
+    assert!(!receiver.is_empty());
+    assert_eq!(receiver.recv().await, Some(1));
+    assert_eq!(receiver.recv().await, Some(2));
+    assert_eq!(receiver.recv().await, None);
+    assert_eq!(receiver.len(), 0);
+    assert!(receiver.is_empty());
+  }
+
+  #[tokio::test(flavor = "current_thread")]
+  async fn receiver_dropped_sending() {
+    let (sender, receiver) = unbounded_channel::<usize>();
+    drop(receiver);
+    let err = sender.send(1).unwrap_err();
+    assert_eq!(err.0, 1);
+  }
+
+  #[tokio::test(flavor = "current_thread")]
+  async fn receiver_recv_then_drop_sender() {
```

Comment:
Can we add some more multi-task tests here?

 - one test with one rx that receives 1000 items, and two senders sending 500 each.
 - the above test, but the receiver sleeping for .1ms between recv calls
 - the above test but the senders sleeping for .1ms between send calls
 - the above test, but 1000 sender tasks instead 

I prefer `rstest` for things like this to ensure we can get good coverage, but duplicating the test would also work.

I don't think you'll see any failures here, but this could probably be heavily used in Deno and it would be ideal to have a good test suite in case someone comes in later and makes changes for some reason.

Reply:
Good idea. I updated it with a loop that goes through all the combinations of that.

Let me know if you can think of any additional tests we should have.